src/main/scala/mesosphere/marathon/api/forwarder/AsyncUrlConnectionRequestForwarder.scala

```diff
@@ -0,0 +1,210 @@
+    Option(request.getHeader("User-Agent")).foreach { ua =>
+      headers += akka.http.scaladsl.model.headers.`User-Agent`(ua)
+    }
+    headers.result()
+  }
+
+  private def createAndConfigureConnection(asyncContext: AsyncContext, url: URL, clientRequest: HttpServletRequest): Future[HttpResponse] = {
+    val method = HttpMethods.getForKey(clientRequest.getMethod).getOrElse(HttpMethod.custom(clientRequest.getMethod))
+    val uri = Uri(url.toString)
+    val proxyHeaders = proxiedRequestHeaders(clientRequest)
+
+    val contentType = Option(clientRequest.getContentType).map(ContentType.parse) match {
+      case Some(Left(ex)) =>
+        logger.error(s"Ignoring unparseable Content-Type: ${clientRequest.getContentType}", ex)
+        ContentTypes.NoContentType
+      case Some(Right(contentType)) =>
+        contentType
+      case None =>
+        ContentTypes.NoContentType
+    }
+
+    val entity = clientRequest.getContentLengthLong match {
+      case 0 =>
+        HttpEntity.empty(contentType)
+      case -1 =>
+        HttpEntity.apply(contentType, Source.empty)
+      case contentLength =>
+        HttpEntity.apply(contentType, contentLength, ServletInputStreamSource.forAsyncContext(asyncContext))
+    }
+
+    val proxyRequest = HttpRequest(method, uri = uri, headers = proxyHeaders, entity = entity)
+    Http().singleRequest(request = proxyRequest, connectionContext = connectionContext, settings = poolSettings)
+  }
+
+  override def forward(url: URL, request: HttpServletRequest, response: HttpServletResponse): Unit = {
+    require(request.isAsyncSupported(), "ServletRequest does not support async mode")
+
+    logger.info(s"Proxying request to ${request.getMethod} $url from $myHostPort")
+
+    val asyncContext = request.startAsync()
+    asyncContext.setTimeout(0L) // delegate timeout to stream
+
+    val result: Future[Done] = try {
+      val hasProxyLoop: Boolean = Option(request.getHeaders(HEADER_VIA)).exists(_.seq.contains(viaValue))
+
+      if (hasProxyLoop) {
+        logger.error("Prevent proxy cycle, rejecting request")
+        response.sendError(BadGateway.intValue, ERROR_STATUS_LOOP)
+        Future.successful(Done)
+      } else {
+        val leaderRequest = try {
+          createAndConfigureConnection(asyncContext, url, request)
+        } catch {
+          case ex: Throwable =>
+            Future.failed(ex)
+        }
+
+        leaderRequest.transformWith {
+          case Failure(ex: akka.stream.StreamTcpException) =>
+            /* Unfortunately, akka-http does not give us a different error message if the TCP connection is established,
+             * but gives no response, or if the TCP connection is refused outright.
+             *
+             * So, we report BadGateway in either case.
+             */
+            logger.error(ERROR_STATUS_CONNECTION_REFUSED, ex)
+            response.sendError(BadGateway.intValue, ERROR_STATUS_CONNECTION_REFUSED)
+            Future.successful(Done)
```

Comment:
That's funny. Here you actually already reply. I don't think there is a need for `Future.successful(Done)`. Also, why not use async/await here?

Reply:
We reply and return success.

Here is why I don't use async/await:

```
[info] Formatting 1 Scala source ProjectRef(uri("file:/Users/tim/src/m8e/marathon/"), "marathon")(compile) ...
[info] Compiling 1 Scala source to /Users/tim/src/m8e/marathon/target/scala-2.12/classes ...
[error] /Users/tim/src/m8e/marathon/src/main/scala/mesosphere/marathon/api/forwarder/AsyncUrlConnectionRequestForwarder.scala:167:35: await must not be used under a try/catch.
[error]           val leaderRequest = Try(await(createAndConfigureConnection(asyncContext, url, request)))
[error]                                   ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 2 s, completed May 20, 2018 11:54:37 PM
marathon(tharper/async-proxy)> 
```